soda/scientific/soda/scientific/distribution/generate_dro.py

```diff
@@ -33,11 +71,48 @@ def generate_continuous_dro(self) -> DRO:
             data = data[~np.isnan(data)]
             logging.warning(
                 f"""{none_count} out of {data_len} rows ({none_count/data_len}%)
-                has None values! To estimate the weights and bins, the null values
-                has been ignored!
+has None values! To estimate the weights and bins, the null values
+has been ignored!
             """
             )
-        weights, bins = np.histogram(data, bins="auto", density=False)
+
+        n_bins = self._compute_n_bins(data)
+
+        # First check whether we have n_bins having lower than data size
+        if n_bins < min(self.maximum_allowed_bin_size, data.size):
+            weights, bins = np.histogram(data, bins="auto", density=False)
+
+        # If we have too much n_bins, then we apply several methods
+        else:
+            # First remove the outliers from data and recompute number of bins
+            outlier_filtered_data = self._remove_outliers_with_iqr(data)
+            n_bins = self._compute_n_bins(outlier_filtered_data)
+
+            # If n_bins is lower than data size then run auto mode again
+            if n_bins < min(self.maximum_allowed_bin_size, data.size):
+                weights, bins = np.histogram(outlier_filtered_data, bins="auto", density=False)
+
+            # If not then take the sqrt of data size and make it as our new n_bins
+            else:
+                n_sqrt_bins = int(np.ceil(math.sqrt(outlier_filtered_data.size)))
+                if n_sqrt_bins < self.maximum_allowed_bin_size:
+                    logging.warning(
+                        f"""Filtering out outliers does not solved the memory error, so we will
+take the square root of the data size to set the number of bins where;
+
+n_bins = {n_bins}
+sqrt(n_bins) = {n_sqrt_bins}
```

Comment:
I'm not entirely sure whether the square root is helpful to users here what do you think?

Reply:
true I'll delete it

Reply:
Also, just for my own understanding and may be this can be added as a comment if true, the rational behind sqrt is this:

1. data size >= `1e12`: sqrt potentially brings down the bins to <= `1e6`
2. data size <= `1e6` for example `999999`: This data size cannot exceed `1e6` bins even with auto so that's safe 
3. data size = `2e6`: `auto` can give a value > `1e6` and sqrt brings down the number of bins to `1414` which is quite drastic compared to the default `1e6` limit

In a case like 3, wouldn't a strategy like exponential backoff be better to get close to the max limit instead of getting to 0.14% of the potential bin size?

Reply:
Hi @mrafayaleem,

Thanks for your feedback and follow-up @mrafayaleem :) 

So, we first check whether the number of bins is greater than the data size. It might be the case like the following array [0.5, 0.43, 0.54, 0.2, 0.1, 0.3 and 1000000000000] generates a super large number of bins as you find out. 

If this is the case, we apply IQR, filter outlier values, and recompute the number of bins. If it still exceeds the data size, then we take the sqrt of data size to use as the number of bins (default strategy of excel to find the number of bins).

If sqrt(data size) still exceeds 1e6, we just wanted to limit an upper boundary as 1e6 to prevent memory issues. (While writing this, I discover a small corner case in the PR so will fix it 👍 )

For your suggestion, exponential backoff seems to be an easy and effective solution as well. We will definitly consider it with the team.